pkg/sql/instrumentation.go

```diff
@@ -470,20 +470,20 @@ func (ih *instrumentationHelper) Setup(
 		}
 	}
 
-	ih.collectExecStats = collectTxnExecStats
-
-	// Don't collect it if Stats Collection is disabled. If it is disabled the
-	// stats are not stored, so it always returns false for previouslySampled.
-	// TODO(117690): Unify StmtStatsEnable and TxnStatsEnable into a single cluster setting.
-	if !collectTxnExecStats && (!previouslySampled && sqlstats.StmtStatsEnable.Get(&cfg.Settings.SV)) {
-		// We don't collect the execution stats for statements in this txn, but
-		// this is the first time we see this statement ever, so we'll collect
-		// its execution stats anyway (unless the user disabled txn stats
-		// collection entirely).
-		statsCollectionDisabled := collectTxnStatsSampleRate.Get(&cfg.Settings.SV) == 0
-		ih.collectExecStats = !statsCollectionDisabled
+	shouldSampleFirstEncounter := func() bool {
+		// If this is the first time we see this statement in the current stats
+		// container, we'll collect its execution stats anyway (unless the user
+		// disabled txn or stmt stats collection entirely). However, we don't want
+		// to collect the stats if the stats container is full, since previouslySampled
+		// will always return false for statements not already in the container.
+		statsCollectionEnabled := collectTxnStatsSampleRate.Get(&cfg.Settings.SV) > 0
+		limitReached := statsCollector.StatementsContainerFull()
```

Comment:
I'm a little confused about how `limitReached` is supposed to affect `ih.collectExecStats`. If we're at the limit shouldn't we override `collectTxnExecStats` if it's set to `true`? Or is the assumption that it's set to true only if it's a collected fingerprint already?

Reply:
There's no assumption of that sort here. We're using `limitReached` specifically just to correct the case where we want to override `collectTxnExecStats` because the container hasn't recorded this fingerprint yet since if it's full it'll never insert this new fingerprint. `collectTxnExecStats` could be set for a statement that's in or not in the container.

If we'd prefer to only set `collectTxnExecStats` for statements in the container, we can investigate the implications of that. The only reason I didn't pursue that is I'm not sure if this sampling logic could be used outside of sql stats for any other reporting / metrics.